rfcs/0045-zkapp-balance-data-in-archive.md

```diff
@@ -0,0 +1,253 @@
+replaying each transaction. With the proposed changes, it can perform those
+verifications after each block.
+
+## Changes to Rosetta
+
+Currently, the `account` endpoint in the Rosetta implementation
+makes at least one SQL query to find a balance and a nonce. If the initial
+query fails, a fallback query is made using two subqueries: at a given block
+height, the balance comes from the most recent transaction (user or
+internal command), while the nonce comes from the most recent user
+command. With the changes here, a single query will always return both the
+balance and nonce, because both items always appear in the `accounts` table.
+
+We'll still need to distinguish canonical from pending block heights,
+and in the latter case, use the Postgresql recursion mechanism to find
+a path back to a canonical block.
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+Storing complete account information for each block likely increases
+the database storage requirements, even though per-transaction
+balances will be removed.
+
+Because balances and nonces are stored per-block, rather than
+per-transaction, related bugs may be more difficult to difficult to
+rectify, when flagged by the replayer.
+
+## Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+A simpler alternative would be to use the existing `balances` table,
+but without the sequence information, rather than storing all the
+account information. That would be enough to allow the replayer and
+Rosetta to work after zkApps are running on a network.  But that would
+limit some possible use cases for the archive data. For example, with
+the complete account information, it's possible to construct the
+ledger corresponding to any block, without running the replayer.
+(We'd still want to the replayer available, to verify ledger hashes, balances,
+and nonces in the archive database.)
+
+## Prior art
+[prior-art]: #prior-art
+
+The proposal here is an extension of existing mechanisms used to store
+and query transaction data in an archive database.
+
+## Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+There are some minor questions posed in the text above. Besides those:
+
+ - For fee transfers and coinbases, the `Applied` status contains
+   `Auxiliary_data.empty`, so no account creation fee information is
+   provided to the archive processor, which uses a calculational hack
+   to determine such fees. Are there obstacles to obtaining those
+   fees on the daemon side?
+
+ - The new database schema has information not present in the current
+   schema, such as the `accounts` table.  It won't be possible to
+   write a simple migration program, that simply shuffles around
+   existing information, if migration is needed. It may be possible to
+   do a migration by modifying the replayer, which maintains a ledger,
+   to write out account information. Will the archive database at the
+   hard fork start empty, or will it contain pre-fork data?
```

Comment:
If we are removing the balances table then we'll need to migrate all the account data at hard fork for existing systems to still be able to refer to historical data.

Reply:
So, it will contain pre-fork data, and not just the ledger at the fork?